components/_impl/workers/subprocess_rpc.py

```diff
@@ -0,0 +1,488 @@
+"""Utilities to handle communication between parent worker.
+
+This module implements three principle facilities:
+    1) Raw IPC (via the Pipe class)
+    2) Exception propagation (via the SerializedException class)
+    3) A run loop for the worker (via the run_loop function)
+"""
+import contextlib
+import dataclasses
+import datetime
+import io
+import marshal
+import os
+import pickle
+import struct
+import sys
+import textwrap
+import threading
+import time
+import traceback
+import types
+import typing
+
+
+# Shared static values / namespace between worker and parent
+BOOTSTRAP_IMPORT_SUCCESS = b"BOOTSTRAP_IMPORT_SUCCESS"
+BOOTSTRAP_INPUT_LOOP_SUCCESS = b"BOOTSTRAP_INPUT_LOOP_SUCCESS"
+WORKER_IMPL_NAMESPACE = "__worker_impl_namespace"
+
+# Constants for passing to and from pipes
+_CHECK = b"\x00\x00"
+_TIMEOUT = b"\x01\x01"
+assert len(_CHECK) == len(_TIMEOUT)
+
+_ULL = "Q"  # Unsigned long long
+_ULL_SIZE = len(struct.pack(_ULL, 0))
+assert _ULL_SIZE == 8
+
+# Text encoding for input commands.
+ENCODING = "utf-8"
+SUCCESS = "SUCCESS"
+
+# In Python, `sys.exit()` is a soft exit. It throws a SystemExit, and only
+# exits if that is not caught. `os._exit()` is not catchable, and is suitable
+# for cases where we really, really need to exit. This is of particular
+# importance because the worker run loop does its very best to swallow
+# exceptions.
+HARD_EXIT = "import os\nos._exit(0)".encode(ENCODING)
+
+# Precompute serialized normal return values
+EMPTY_RESULT = marshal.dumps({})
+SUCCESS_BYTES = marshal.dumps(SUCCESS)
+
+
+# =============================================================================
+# == Raw Communication ========================================================
+# =============================================================================
+
+# Windows does not allow subprocesses to inherit file descriptors, so instead
+# we have to go the the OS and get get the handle for the backing resource.
+IS_WINDOWS = sys.platform == "win32"
+if IS_WINDOWS:
+    import msvcrt
+    def to_handle(fd: typing.Optional[int]) -> typing.Optional[int]:
+        return None if fd is None else msvcrt.get_osfhandle(fd)
+
+    def from_handle(handle: typing.Optional[int], mode: int) -> typing.Optional[int]:
+        return None if handle is None else msvcrt.open_osfhandle(handle, mode)
+
+else:
+    to_handle = lambda fd: fd
+    from_handle = lambda fd, _: fd
+
+
+class _TimeoutPIPE:
```

Comment:
This is somewhat complicated, I suspect if you wanted you could punt most of this complexity to using the thread safe queue.Queue; the singleton becomes a single consumer and you can multiple producers. But the solution you have here seems to work, and seems to have some extra features like deduplication.

Reply:
I experimented with putting the read on a side thread, and then timing out communication with that thread. (Am I understanding correctly that that's the gist of what you're getting at?) The issue was that it's horribly slow. (About 1000x slower than keeping the `os.read` on the main thread, and unacceptably high based on my discussions with Eric.)